src/main/java/nextstep/courses/domain/SessionEnrollmentContext.java

```diff
@@ -2,77 +2,114 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class SessionEnrollmentContext {
     private final List<Student> students = new ArrayList<>();
-    private Status progressStatus = Status.NOT_STARTED;
+    private SessionStatus progressSessionStatus = SessionStatus.NOT_STARTED;
+    private EnrollmentStatus enrollmentStatus = EnrollmentStatus.ENROLLABLE;
     private long maxEnrollment;
 
     public SessionEnrollmentContext(long maxEnrollment) {
         this.maxEnrollment = maxEnrollment;
     }
 
-    public SessionEnrollmentContext(long maxEnrollment, Status progressStatus) {
+    public SessionEnrollmentContext(long maxEnrollment, SessionStatus progressSessionStatus) {
         this(maxEnrollment);
-        this.progressStatus = progressStatus;
+        this.progressSessionStatus = progressSessionStatus;
     }
 
-    public SessionEnrollmentContext(long maxEnrollment, Status progressStatus, List<Student> students) {
-        this(maxEnrollment, progressStatus);
+    public SessionEnrollmentContext(long maxEnrollment, SessionStatus progressSessionStatus, EnrollmentStatus enrollmentStatus, List<Student> students) {
+        this(maxEnrollment, progressSessionStatus);
+        this.enrollmentStatus = enrollmentStatus;
         this.students.addAll(students);
     }
 
-    public boolean statusEquals(Status status) {
-        return progressStatus == status;
+    public boolean statusEquals(SessionStatus sessionStatus) {
+        return progressSessionStatus == sessionStatus;
     }
 
     public long getNumberOfStudents() {
-        return this.students.size();
+        return getEnrolledStudents().size();
     }
 
     public boolean isEnrollable() {
-        return this.progressStatus == Status.IN_PROGRESS &&
-                this.maxEnrollment > students.size();
+        return isNotFinished() && isNotFull() && isStatusEnrollable();
+    }
+
+    private List<Student> getEnrolledStudents() {
+        return this.students.stream()
+                .filter(Student::isEnrolled)
+                .collect(Collectors.toList());
+    }
+
+    private boolean isNotFinished() {
+        return this.progressSessionStatus != SessionStatus.FINISHED;
```

Comment:
```suggestion
        return progressSessionStatus.isNotFinished();
```
enum이 jvm 상에서 인스턴스 하나를 보장하는 것을 제외하면 class와 성질이 같다.
enum에도 위와 같이 메시지를 보내보면 어떨까?